metrics/src/main/scala/com/evolutiongaming/kafka/flow/TopicFlowMetrics.scala

```diff
@@ -50,7 +51,7 @@ object TopicFlowMetrics {
   implicit def topicFlowOfMetricsOf[F[_]: Monad: MeasureDuration]: MetricsOf[F, TopicFlowOf[F]] =
     topicFlowMetricsOf[F] transform { implicit metrics => topicFlowOf =>
       new TopicFlowOf[F] {
-        def apply(consumer: Consumer[F], topic: Topic) =
+        def apply(consumer: Consumer[F, String, ByteVector], topic: Topic) =
```

Comment:
The need to drag `[String, ByteVector]` everywhere could be another reason why we might want to have our own `Consumer`.

Reply:
We can define `type Consumer[F[_]] = Consumer[F, String, ByteVector]` somewhere to avoid it. But I don't see an issue here, it's not more verbose that anything else.

Reply:
So you are saying that writing `Consumer[F, String, ByteVector]` in 24 places is not more verbose than writing `Consumer[F]`?

Reply:
I'm saying that using `type Consumer[F[_]] = Consumer[F, String, ByteVector]` makes it looks the same but without additional abstraction w/ logic, Of factory and misleading between skafka and flow consumers.

Reply:
I think `Of` factory question is not anyhow related to this topic. It is a different discussion. You can have `Of` with `skafka` consumer or with our own with exactly the same success.

It is `Consumer[F]` vs `Consumer[F, String, ByteVector]`.

I see that the current pull request in this area:
(1) changes `Consumer[F]` to `Consumer[F, String, ByteVector]`,
(2) creates a large `TestConsumer` class with most methods failing,
(3) allows the one to accidentally call the much larger set of methods on underlying consumer.

Does it really simplify or make the code straightforward in this specific area? It creates a lot of lines of the code and the whole additional area to fail in. I agree `type` resolves (3), but it does not resolve other two issues.

It is a good programming practice to use as less power as we can. This is one of the reasons to use Tagless Final in the first place. Having our own very limited `Consumer` serves exactly this purpose - we limit ourselves to a subset of the functionality `skafka` provides getting more safety and simplifying the reasoning and the tests.

I agree about misleading name, i.e. the same as in `skafka`. I am not sure what the right policy could be. Naming things is hard. I think we were following naming convention of `kafka-journal`, i.e. to give exactly the same name for the things we want to be used instead. We could think about another policy and name it differently, i.e. `ByteConsumer` or something like that. I am not sure it is better. What do you think? :thinking:

Reply:
> (1) changes Consumer[F] to Consumer[F, String, ByteVector]

Please check latest commit, that's not true anymore.

> creates a large TestConsumer class with most methods failing

I prefer to have less production code & simple implementation with some additional test logic.

> allows the one to accidentally call the much larger set of methods on underlying consumer.

_one_ - kafka-flow library, probably? There are no user interaction w/ end user API as far as I see.

Additional abstraction makes already overcomplicated code even more complicated. The change has purpose to simplify codebase and make readability easier and cheaper to maintain.

I agree that test implementation is huge and we should come out with better solution (might be unpopular in FP world mocks?), but should not stop us from making production code simpler.

Reply:
> Additional abstraction makes already overcomplicated code even more complicated. The change has purpose to simplify codebase and make readability easier and cheaper to maintain.
> ?
> I agree that test implementation is huge and we should come out with better solution (might be unpopular in FP world mocks?), but should not stop us from making production code simpler.

From my subjective opinion this change makes codebase more complicated and less readable :shrug:

This library, currently, uses traits extensively to simplify the testing. It also use a unit testing approach based on Tagless Final (https://medium.com/wix-engineering/functional-testing-with-tagless-final-50eeacf5df6). From my opinion such approach allows the one to write the code which is shorter and more reliably that the one based on mocks. I will be happy to teach you the one in person.

I do not think mocks are popular among anyone these days regardless if they are FP, OOP etc. developers. The people tend to prefer stubs for various reasons. I might be mistaken. See https://martinfowler.com/articles/mocksArentStubs.html for more details on the topic.